### PR TITLE
ENH Add beta_fragility_heuristic and gpd_risk_estimates functions

### PR DESCRIPTION
**Pull request by jeyoor**
Hi,

I thought it'd be interesting to add some risk metrics to empyrical for use in zipline/pyfolio.

Here's the papers I used.

   * Beta Fragility: https://www.imf.org/external/pubs/ft/wp/2012/wp12216.pdf

   * GPD Risk Estimates: https://link.springer.com/article/10.1007/s10614-006-9025-7

I wasn't sure about whether or not docstrings and tests are required for subroutines. (I'm missing some docstrings and tests for subroutines of gpd_risk_estimates.)

If the above are required, please feel free to ping me back and I'll work on adding them.

Thanks!

Jeyan

### CODE REVIEW

**Comment by jeyoor** [comment]
Ah looks like I forgot to check flake.

Will push a patchset.

**Comment by twiecki** [comment]
Thanks @jeyoor, this looks great. I'll try to take a look next week (traveling).

**Comment by twiecki** [comment]
Sorry @jeyoor, I completely dropped the ball on this. This looks great, if you solve the merge conflicts we can go ahead and merge this.

**Comment by jeyoor** [comment]
Hi @twiecki,

No worries; schedule pressure can be tough.

I solved the merge conflicts, but tests are currently failing.
I am planning to look into this some more when I have some time.

Tests run fine on quantopian:master so it must be something I added or changed, but it isn't clear to me right now exactly what's causing the tests to fail...

I'll keep you posted on the status.

Thanks!

**Comment by jeyoor** [comment]
Hi @twiecki,

I think everything passes now.

I wasn't sure if the new functions should be in SIMPLE_STAT_FUNCS or FACTOR_STAT_FUNCS,  but they seemed to fit best in FACTOR_STAT_FUNCS to I moved them there.

I look forward to your feedback.

Thanks!

File: empyrical/__init__.py.orig [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#
```
Comment (twiecki) [inline empyrical/__init__.py.orig]: This file shouldn't show up.

File: empyrical/stats.py.orig [hunk 2]
```diff
@@ -0,0 +1,1616 @@
+#
```
Comment (twiecki) [inline empyrical/stats.py.orig]: this file neither

**Comment by twiecki** [comment]
Thanks for picking this back up! Looks like the merge wasn't quite successful yet. If everything fails, you could from a clean master you cherry-pick your commits, or recreate those functions.

**Comment by jeyoor** [comment]
Doh! Looks like I accidentally added some merge backup files.
How does it look now?

**Comment by jeyoor** [comment]
ah missed one file... working on it

**Comment by jeyoor** [comment]
Hi @twiecki,

How are the diffs looking after the most recent update?

Thanks!

File: empyrical/stats.py [hunk 3]
```diff
@@ -993,6 +995,325 @@ def cagr(returns, period=DAILY, annualization=None):
...
 
+def beta_fragility_heuristic(returns, factor_returns):
+    """
+    Estimate fragility to drops in beta.
+    A negative return value indicates potential losses
+    could follow volatility in beta.
+    The magnitude of the negative value indicates the size of
+    the potential loss.
+
+    seealso::
```
Comment (twiecki) [inline empyrical/stats.py]: I would put this under `Notes` (with `---` underlining) at the bottom.

File: empyrical/stats.py [hunk 4]
```diff
@@ -993,6 +995,325 @@ def cagr(returns, period=DAILY, annualization=None):
     return ending_value ** (1. / no_years) - 1
 
 
+def beta_fragility_heuristic(returns, factor_returns):
+    """
+    Estimate fragility to drops in beta.
```
Comment (twiecki) [inline empyrical/stats.py]: these line-breaks are a bit odd.

File: empyrical/stats.py [hunk 5]
```diff
@@ -993,6 +995,325 @@ def cagr(returns, period=DAILY, annualization=None):
...
+
+    return beta_fragility_heuristic_aligned(
+        *_aligned_series(returns, factor_returns))
+
+
+def beta_fragility_heuristic_aligned(returns, factor_returns):
+    """
+    Estimate fragility to drops in beta
+
+    seealso::
```
Comment (twiecki) [inline empyrical/stats.py]: same.

**Comment by twiecki** [comment]
Looks good, I think I left some comments on an earlier commit related to formatting.

**Comment by gusgordon** [comment]
Closing this as it's old. Feel free to reopen if this is still relevant.

**Comment by jeyoor** [comment]
@gusgordon ,
Thank you for the reminder on this!
I'm a bit slammed at the moment, but once I get some free time I plan to update the branch per the comments and ping you back.
Thanks again!
Jey

**Comment by jeyoor** [comment]
Moved to new pull request here https://github.com/quantopian/empyrical/pull/118

**Pull request closed without merging**